### ISSUE TITLE
update & consent scopes on subsequent requests

### ISSUE THREAD

**Issue by KtorZ** [panva/node-oidc-provider#164]
Hi, I am trying to find a way to retrieve the scopes previously consented by a user upon a new authorization request. Here's a simple scenario explaining the issue:

1 - A non authenticated end-user is redirected to the authorization endpoint
2 - oidc-provider delegates the interaction to my server 
3 - The interaction completes with `{ login: { ... }, consent: { scope: 'openid my:scope' } }`
[ ... later ... ]
4 - The same end-user is redirected to the authorization endpoint with `scope=openid my:other:scope`

At that point, in a simple flow (no particular prompt, user already authenticated with an active and valid session in the AS), the `interactionCheck` function get called to see if any kind of interaction is needed.
In this case, I would like to reply that a `consent_required` interaction is needed since a new scope `my:other:scope` is requested and wasn't consented in the first request. 

However, I can't find a way to know from the oidc `Session` what scope was consented when the client was first authorized. Do I have to keep track of these authorized scopes for each client? To me, it sounds like something that `oidc-provider` should be doing but I might get the process wrong.

Thanks.

**Comment by panva** [comment]
You're looking for the [`interactionCheck`](https://github.com/panva/node-oidc-provider/blob/master/lib/helpers/defaults.js#L348-L375) configuration option to specify this logic. You can return your consent_required again in that one. How you store and retrieve what has been consented to in the past is up to you, you may use the session object, custom cookie or even the your account object, whatever you have at your disposal.

**Comment by KtorZ** [comment]
Yup, so I got the flow correctly. After the first authorization, the openid session looks roughly like this (using a MongoDB adapter):

```json
{
	"_id" : "93fe397b-1782-4534-b144-c64f789b67e0",
	"expiresAt" : ISODate("2018-10-26T21:17:10.415Z"),
	"account" : "479531f6-765a-4632-b5d0-6a47ece78227",
	"loginTs" : 1509029850,
	"authorizations" : {
		"myClient" : {
			"sid" : "e5852ab2-80a8-4f9b-9ccd-b1e0d539fded"
		}
	}
}
```

Wouldn't it make sense for the `oidc-provider` to also store consented scope next to the `sid` for each authorized client? Or do you consider this as a task my server should be managing (and therefore, keeps tracks of authorized clients and sessions myself which is actually part of the reason why I use `oidc-provider` in a first place).

**Comment by panva** [comment]
The interaction result `consent.scope` is only used to reduce the scopes from the request that get then assigned to the code/tokens, it is not used to keep track of what's accepted. I do consider these policies to be a part of each AS and hence not something coming out of the box so long as there's a way for you to achieve this policy, which I believe there is.

**Comment by panva** [comment]
That being said i agree there is a place for this in the package, can you provide a working PR base to iterate on?

I'm thinking the param or result scopes get also assigned next to `sid`, then it will be simple for you to you extend the interactionCheck function and check the scope difference between session and request.

**Comment by KtorZ** [comment]
> I'm thinking the param or result scopes get also assigned next to sid, then it will be simple for you to you extend the interactionCheck function and check the scope difference between session and request.

This is actually exactly what I have in mind. It is already quite handy to retrieve an active session at that moment but the information available in a session is quite poor. 
I'll work tomorrow on a possible extension then and submit it as a PR :+1:

**Comment by panva** [comment]
well yeah, but all it takes is to provide the interactionCheck function, which you'll have to anyway. i just confirmed working with this function is all it takes.

**Comment by panva** [comment]
PR merged, to be released with 2.5.0 when #167 is also resolved.
